features/nvstore/source/nvstore_int_flash_wrapper.cpp

```diff
@@ -0,0 +1,184 @@
+
+static mbed::FlashIAP flash;
+
+static size_t get_page_size(void)
+{
+    return MIN(flash.get_page_size(), MAX_PAGE_SIZE);
+}
+
+static int program_flash(size_t size, uint32_t address, const uint8_t *buffer)
+{
+    int ret;
+    uint32_t page_size, aligned_size, remainder;
+    uint8_t rem_buf[MAX_PAGE_SIZE];
+
+    page_size = get_page_size();
+    remainder = size % page_size;
+    aligned_size = size - remainder;
+    if (aligned_size) {
+        ret = flash.program(buffer, address, aligned_size);
+
+        if (ret) {
+            return -1;
+        }
+        address += aligned_size;
+        buffer += aligned_size;
+    }
+
+    if (!remainder) {
+        return 0;
+    }
+
+    memset(rem_buf, NVSTORE_BLANK_FLASH_VAL, page_size);
+    memcpy(rem_buf, buffer, remainder);
+    ret = flash.program(buffer, address, page_size);
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+size_t nvstore_int_flash_get_sector_size(uint32_t address)
+{
+    return flash.get_sector_size(address);
+}
+
+size_t nvstore_int_flash_get_flash_start()
+{
+    return flash.get_flash_start();
+}
+
+size_t nvstore_int_flash_get_flash_size()
+{
+    return flash.get_flash_size();
+}
+
+int nvstore_int_flash_init(void)
+{
+    int ret;
+
+    ret = flash.init();
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+int nvstore_int_flash_deinit(void)
+{
+    int ret;
+
+    ret = flash.deinit();
+    if (ret) {
+        return -1;
+    }
+
+
+    return 0;
+}
+
+int nvstore_int_flash_read(size_t size, uint32_t address, uint32_t *buffer)
+{
+    int ret;
+
+    if (!buffer) {
+        return -1;
+    }
+
+    if (!size)
+    {
+        return -1;
+    }
+    ret = flash.read(buffer, address, size);
+
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+
+int nvstore_int_flash_erase(uint32_t address, size_t size)
+{
+    int ret;
+    size_t sector_size = nvstore_int_flash_get_sector_size(address);
+
+    if ((!size) || (size % sector_size) || (address % sector_size)) {
+        return -1;
+    }
+
+    // No need to iterate over sectors - Flash driver's erase API does it for us
+    ret = flash.erase(address, sector_size);
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+
+int nvstore_int_flash_write(size_t size, uint32_t address, const uint32_t *buffer)
+{
+    int ret;
+    uint32_t page_size = get_page_size();
+    uint32_t sector_size = nvstore_int_flash_get_sector_size(address);
+    uint32_t chunk;
+    uint8_t *buf = (uint8_t *) buffer;
+
+    if ((!size) || (!buffer) || (address % page_size)) {
+        return -1;
+    }
+
+    while (size) {
+        chunk = MIN(sector_size - (address % sector_size), size);
+        ret = program_flash(chunk, address, buf);
+        if (ret) {
+            return -1;
+        }
+        size -= chunk;
+        address += chunk;
+        buf += chunk;
+        sector_size = nvstore_int_flash_get_sector_size(address);
+    }
+
+    return 0;
+}
+
+
+#endif
```

Comment:
This whole file makes it seem like FlashIAP is not very useful. Are these additional checks that are needed? Should this just be a private part of NVStore and not its own wrapper file?

Reply:
This can be moved inside NVStore. However, please note the extra logics in the program function. Right now, FlashIAP is very limited if we want to program data crossing sectors or if not on aligned page sizes. The nvstore_int_flash_write and program_flash functions take care of these limitations. I believe FlashIAP should handle this, but if not - that extra logics is required. Other functions (erase, read, get sector size etc.) don't need that extra logics indeed.

Reply:
>  I believe FlashIAP should handle this

@davidsaada We shall move to `FlashIAP`. Can you send PR to move them there? If this is a functionality that flash API should cover, and its not currently covering.

For example `nvstore_int_flash_erase` can be eliminated here, that check  could be added to the flash erase sector implementation.

Reply:
0xc0170 - I believe this should be there (don't think the caller should worry about page sizes and sector boundaries), but maybe others would disagree with me. @sg- , @geky - what do you think?

Reply:
We write APIs to be useful. If we've missed something, lets add it. I cant think of a good reason not to.

Reply:
Right. Agree with that. So this was indeed modified in a [separate PR](https://github.com/ARMmbed/mbed-os/pull/6140). Got a derived branch that eliminates the flash wrapper and uses FlasIAP directly, which will be applied once the separate PR is merged.

Reply:
👍